Compfiles/Imo1970P4.lean

```diff
@@ -21,15 +20,831 @@ equals the product of the nubmers in the other set.
+
+lemma n_eq_1_of_contains_one
+  (n : ℕ) (n_not_zero : n ≠ 0) (contains_one : 1 ∈ Finset.Icc n (n + 5)) : n = 1 := by
+  simp_all only [ne_eq, Finset.mem_Icc, le_add_iff_nonneg_left, zero_le, and_true]
+  omega
+
+lemma diffs_of_disjoint (t s w : Finset ℕ) (t_subst_s : t ⊆ s) (disjoint: Disjoint t w) : t ⊆ s \ w := by
+  simp [Finset.subset_sdiff, *]
+
+lemma not_empty_subst_of_nonempty (t : Finset ℕ) (t_nonempty : t.Nonempty) : ¬ t ⊆ ∅ := by
+  rw [Finset.subset_empty]
+  exact t_nonempty.ne_empty
+
+lemma subset_of_union_right (t s : Finset ℕ) : t ⊆ s ∪ t := by
+  have := @Finset.subset_union_right ℕ _ s t
+  exact this
+
+lemma mem_of_subst (k : ℕ) (t s : Finset ℕ) (k_in_t : k ∈ t) (t_subst_s : t ⊆ s) : k ∈ s := by
+  apply t_subst_s
+  simp_all only
+
+lemma contradiction_of_finset_icc_1_6 (s1 s2 : Finset ℕ) (partition : s1 ∪ s2 = Finset.Icc 1 6)
+  (disjoint : s1 ∩ s2 = ∅) (eq_prod: ∏ m ∈ s1, m = ∏ m ∈ s2, m) : False := by
+  have : 5 ∈ s1 ∪ s2 := by
+    rw[partition]
+    simp
+  simp at this
+  cases this
+  · case inl five_in_s1 =>
+    have s2_in_s1_s2: s2 ⊆ s1 ∪ s2 := subset_of_union_right s2 s1
+    have explicit_s2 : s2 ⊆ {1, 2, 3, 4, 6} := by
+      have five_not_in_s2 : Disjoint s2 {5} := by
+        have s1_s2_disjoint : Disjoint s1 s2 := Finset.disjoint_iff_inter_eq_empty.mpr disjoint
+        simp_all only [Finset.disjoint_singleton_right]
+        intro five_in_s2
+        dsimp[Disjoint] at s1_s2_disjoint
+        have five_set_in_s1 : {5} ⊆ s1 := by
+          simp_all only [Finset.singleton_subset_iff]
+        have five_set_in_s2 : {5} ⊆ s2 := by
+          simp_all only [Finset.singleton_subset_iff]
+        have set_five_in_empty := s1_s2_disjoint five_set_in_s1 five_set_in_s2
+        have : Nonempty ({5} : Finset ℕ)  := by
+          simp_all only [Finset.singleton_subset_iff, Finset.mem_singleton, nonempty_subtype, exists_eq]
+        have : ({5} : Finset ℕ).Nonempty := by
+          simp_all only [Finset.singleton_subset_iff, Finset.mem_singleton, nonempty_subtype, exists_eq, Finset.singleton_nonempty]
+        apply not_empty_subst_of_nonempty {5} this
+        exact set_five_in_empty
+      have s2_in_interval : s2 ⊆ Finset.Icc 1 6 := by
+        rw[partition] at s2_in_s1_s2
+        exact s2_in_s1_s2
+      have explicit_interval: Finset.Icc 1 6 = {1, 2, 3, 4, 5, 6} := by
+        rfl
+      rw [explicit_interval] at s2_in_interval
+      have := diffs_of_disjoint s2 (s1 ∪ s2) {5} s2_in_s1_s2 five_not_in_s2
+      rw [partition] at this
+      rw[explicit_interval] at this
+      simp_all only [Finset.disjoint_singleton_right]
+      exact this
+    have others : ∀ k ∈ s2, ¬ 5 ∣ k := by
+      intro k
+      intro k_in_s2
+      have k_in_explicit_s2 : k ∈ ({1, 2, 3, 4, 6} : Finset ℕ) := by
+        exact mem_of_subst k s2 {1, 2, 3, 4, 6} k_in_s2 explicit_s2
+      intro five_div_k
+      simp_all only [Finset.mem_insert, Finset.mem_singleton]
+      cases k_in_explicit_s2 with
```

Comment:
A fancier way to do this is:
```lean
obtain rfl | rfl | rfl | rfl | rfl | rfl := k_in_explicit_s2 <;> contradiction
```

Reply:
What does this code do though? I don't really understand it.

If I did understand it, I think the question for me would be if it's better style than:

```
cases k_in_explicit_s2 with
      | inl h =>
        subst h
        contradiction
      | inr h_1 =>
        cases h_1 with
        | inl h =>
          subst h
          contradiction
        | inr h_2 =>
          cases h_2 with
          | inl h =>
            subst h
            contradiction
          | inr h_1 =>
            cases h_1 with
            | inl h =>
              subst h
              contradiction
            | inr h_2 =>
              subst h_2
              contradiction
```

The latter uses very standard idioms (at the cost of some repetition) and is easier to read than the former.

I've put the golfed version in for now, but personally I tend to prefer code that's easy to read over code that's easy to write.

This might be different in the world of tactic-based theorem provers, as tactic-based programs are infamously difficult to read anyway (I think one pretty much has to read them with compiler in hand).

What do you think David?

Reply:
Ah, oops --- I noticed just now that I put one too many `rfl`s in there. The last one is not needed.

My thought process here was as follows:

First, we can use `obtain` with its pattern syntax to flatten the case analysis:
```lean
      obtain h | h_1 | h_2 | h_1 | h := k_in_explicit_s2
      · subst h
        contradiction
      · subst h_1
        contradiction
      · subst h_2
        contradiction
      · subst h_1
        contradiction
      · subst h
        contradiction
```
Then we note that in all cases we are immediately doing `subst h`. The `rfl` keyword for `obtain` does that automatically. So the above is equivalent to 
```lean
      obtain rfl | rfl | rfl | rfl | rfl := k_in_explicit_s2
      · contradiction
      · contradiction
      · contradiction
      · contradiction
      · contradiction
```

Then we notice that each branch is exactly the same, so we can collapse them with the `<;>` combinator:

```lean
      obtain rfl | rfl | rfl | rfl | rfl := k_in_explicit_s2 <;> contradiction
```

Reply:
Very clever! I've put it in, but again if we golf code like this it might be difficult to read for beginners. (again, not sure if proof readability is as important as readability of software source code).